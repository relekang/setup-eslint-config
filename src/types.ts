import { Linter } from "eslint";
import prompts from "prompts";

export interface ConfigOptions {
  yarn: boolean;
  babel: boolean;
  typescript: boolean;
  flowtype: boolean;
  react: boolean;
  vue: boolean;
  prettier: boolean;
  jest: boolean;
  mocha: boolean;
  ava: boolean;
  cypress: boolean;
  node: boolean;
  dependencies: string[];
}

export interface PackageInfo {
  name: string;
  version: string;
  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
}

export interface SetupConfig {
  name: string;
  prompts: prompts.PromptObject[];
  skipDetectedPrompts?: boolean;
  packageInfo: PackageInfo;
  createEslintConfig: (
    config: Omit<ConfigOptions, "dependencies">
  ) => Linter.Config;
  createDependencyList: (
    config: Omit<ConfigOptions, "dependencies">
  ) => string[];
}
