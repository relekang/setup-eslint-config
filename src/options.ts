import fs from "fs";
import path from "path";
import prompts from "prompts";
import { promisify } from "util";
import { ConfigOptions, PackageInfo, SetupConfig } from "./types";
import { exists } from "./files";
import { dependencyString, hasDependency, oneOf } from "./utils";
import { Linter } from "eslint";

const readFile = promisify(fs.readFile);

export async function buildOptions(
  setupConfig: SetupConfig,
  currentConfig: Linter.Config
): Promise<ConfigOptions> {
  const cwd = process.cwd();
  const packageInfo = JSON.parse(
    (await readFile(path.resolve(cwd, "package.json"))).toString()
  ) as unknown as PackageInfo;
  const detected = {
    yarn: await exists(path.resolve(cwd, "yarn.lock")),
    babel: await oneOf(
      exists(path.resolve(cwd, ".babelrc")),
      exists(path.resolve(cwd, "babel.config.js"))
    ),
    typescript: await oneOf(
      hasDependency(packageInfo, "typescript"),
      exists(path.resolve(cwd, "tsconfig.json"))
    ),
    flowtype: await exists(path.resolve(cwd, ".flowconfig")),
    react: hasDependency(packageInfo, "react"),
    vue: hasDependency(packageInfo, "vue"),
    prettier: hasDependency(packageInfo, "prettier"),
    jest: await oneOf(
      hasDependency(packageInfo, "jest"),
      hasDependency(packageInfo, "@vue/cli-plugin-unit-jest")
    ),
    mocha: hasDependency(packageInfo, "mocha"),
    ava: hasDependency(packageInfo, "ava"),
    cypress: hasDependency(packageInfo, "cypress"),
    node: !!currentConfig?.env?.node,
  };
  const detectedKeys = Object.keys(detected);

  const promptsConfig = setupConfig.skipDetectedPrompts
    ? setupConfig.prompts.map(
        (prompt: prompts.PromptObject): prompts.PromptObject => ({
          ...prompt,
          type: detectedKeys.includes(prompt.name as string)
            ? null
            : prompt.type,
        })
      )
    : setupConfig.prompts;
  const answers = await prompts(promptsConfig);
  const config = {
    ...detected,
    ...answers,

    dependencies: setupConfig
      .createDependencyList({ ...detected, ...answers })
      .map(dependencyString(setupConfig.packageInfo)),
  };

  return config;
}
