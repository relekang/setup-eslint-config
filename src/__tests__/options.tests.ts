import { buildOptions } from "../options";
import { SetupConfig } from "../types";

test("buildOptions with node set in current config", async () => {
  const setupConfig: SetupConfig = {
    name: "test",
    prompts: [],
    createEslintConfig: () => ({}),
    createDependencyList: () => ["eslint"],
    packageInfo: {
      name: "test-package",
      version: "1.42.0",
      devDependencies: { eslint: "^42" },
    },
  };

  expect(
    await buildOptions(setupConfig, { env: { node: true } })
  ).toMatchSnapshot();
});

test("buildOptions without env in current config", async () => {
  const setupConfig: SetupConfig = {
    name: "test",
    prompts: [],
    createEslintConfig: () => ({}),
    createDependencyList: () => ["eslint"],
    packageInfo: {
      name: "test-package",
      version: "1.42.0",
      devDependencies: { eslint: "^42" },
    },
  };

  expect(await buildOptions(setupConfig, {})).toMatchSnapshot();
});
