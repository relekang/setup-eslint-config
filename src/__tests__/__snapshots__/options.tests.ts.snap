// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildOptions with node set in current config 1`] = `
Object {
  "ava": false,
  "babel": false,
  "cypress": false,
  "dependencies": Array [
    "eslint@^42",
  ],
  "flowtype": false,
  "jest": true,
  "mocha": false,
  "node": true,
  "prettier": true,
  "react": false,
  "typescript": true,
  "vue": false,
  "yarn": false,
}
`;

exports[`buildOptions without env in current config 1`] = `
Object {
  "ava": false,
  "babel": false,
  "cypress": false,
  "dependencies": Array [
    "eslint@^42",
  ],
  "flowtype": false,
  "jest": true,
  "mocha": false,
  "node": false,
  "prettier": true,
  "react": false,
  "typescript": true,
  "vue": false,
  "yarn": false,
}
`;
